### PR TITLE
Allow to set the accept input attribute to arbitrary values

### PR DESCRIPTION
**Pull request by mayjs**
This supersedes #750.
It still solves #546 but may also prove to be useful in other scenarios by allowing to set more mediatypes.

### CODE REVIEW

**Comment by svenstaro** [comment]
Pretty neat and clean! I'd almost be happy to accept it as is but I think this feature is complex enough to require another example line of documentation in the README and also I'd like to see a test for this. The test should be rather simple and you can copy from the existing body of tests.

**Comment by mayjs** [comment]
Thanks for the feedback!

I added some tests and a line for the README.
How / when do you update the `USAGE` section in the README?
Is that something you do when a new release is created or should I adjust that part as well?
I was worried about the version number.

Are the tests fine for you or did you have something else in mind?
Thanks for making me write them, that way I learned about `rstest` and how easily you can do fixture-based testing and parametrized tests using it 🙂.

File: src/args.rs [hunk 1]
```diff
@@ -104,6 +111,19 @@ pub struct CliArgs {
...
+    pub media_type: Option<Vec<MediaType>>,
+
+    /// Directly specify the uploadable media type expression
+    #[clap(
+        short = 'M',
+        long = "raw-media-type",
+        requires = "file-upload",
+        conflicts_with = "media-type"
+    )]
+    pub media_type_raw: Option<String>,
```
Comment (svenstaro) [inline src/args.rs:125]: Is this really useful? At the point where you need this level of control, wouldn't it be better to just generally set a header using `--header`?

File: tests/upload_files.rs [hunk 2]
```diff
@@ -167,3 +167,23 @@ fn upload_to_symlink_directory(
...
+#[case(server(&["-u", "-M", "test_value"]), Some("test_value"))]
+fn set_media_type(
+    #[case] server: TestServer,
+    #[case] expected_accept_value: Option<&str>,
+) -> Result<(), Error> {
+    let body = reqwest::blocking::get(server.url())?.error_for_status()?;
+    let parsed = Document::from_read(body)?;
+
+    let input = parsed.find(Attr("id", "file-input")).next().unwrap();
+    assert_eq!(input.attr("accept"), expected_accept_value);
```
Comment (svenstaro) [inline tests/upload_files.rs:186]: Very nice test!

**Comment by svenstaro** [comment]
Don't worry about updating the usage, I will do it with the next release. :)

**Review by svenstaro — CHANGES_REQUESTED** [review]
Almost there! :)

File: src/args.rs (hunk 1 above)
Comment (mayjs) [reply src/args.rs:125]: How would you achieve the same thing using a header?

I think the `raw-media-type` parameter can be useful if you want to specify a filetype or a media type that is not included in the `media-type` parameter.
There are a lot of possibilities for browsers to use this value and offer the right file choosing option to the user, and I think by adding this option we gain a lot of flexibility without adding a lot of complexity.

To give an example, if you wanted to upload a PDF document, you could use `miniserve -u -M application/pdf` and your file chooser would automatically filter for PDF files.
On stock Android, this results in a nice overview of your most recent PDF files all in one place, allowing to quickly find your file.

Comment (svenstaro) [reply src/args.rs:125]: Nevermind, of course this can't be accomplished using header. I really like your explanation. Merging as is!

**Pull request merged**